src/shm/pool/multi.rs

```diff
@@ -0,0 +1,295 @@
+//!         // There's no limit to the amount of buffers we can allocate to our surface but since
+//!         // shm buffers are released fairly fast, it's unlikely we'll need more than double buffering.
+//!         for i in 0..2 {
+//!             self.surface.1 = i;
+//!             if let Some((offset, buffer, slice)) = self.pool.create_buffer(
+//!                 100,
+//!                 100 * 4,
+//!                 100,
+//!                 &self.surface,
+//!                 Format::Argb8888,
+//!                 conn,
+//!             ) {
+//!                 /*
+//!                     insert drawing code here
+//!                 */
+//!                 surface.attach(conn, Some(&buffer), 0, 0);
+//!                 surface.commit(conn);
+//!                 // We exit the function after the draw.
+//!                 return;
+//!             }
+//!         }
+//!         /*
+//!             If there's no buffer available we can for example request a frame callback
+//!             and trigger a redraw when it fires.
+//!             (not shown in this example)
+//!         */
+//!     }
+//! }
+//!
+//! fn draw(slice: &mut [u8]) {
+//!     todo!()
+//! }
+//!
+//! ```
+//!
+
+use std::io;
+
+use std::sync::{
+    atomic::{AtomicBool, Ordering},
+    Arc,
+};
+use wayland_client::{
+    protocol::{wl_buffer, wl_shm, wl_shm_pool},
+    ConnectionHandle, Proxy, WEnum,
+};
+
+use super::raw::RawPool;
+
+/// This pool manages buffers associated with keys.
+/// Only one buffer can be attributed to a given key.
+#[derive(Debug)]
+pub struct MultiPool<K: PartialEq + Clone> {
+    buffer_list: Vec<BufferHandle<K>>,
+    pub(crate) inner: RawPool,
+}
+
+#[derive(Debug)]
+struct BufferHandle<K: PartialEq + Clone> {
+    free: Arc<AtomicBool>,
+    size: usize,
+    used: usize,
+    offset: usize,
+    buffer: Option<wl_buffer::WlBuffer>,
+    key: K,
+}
+
+impl<E: PartialEq + Clone> From<RawPool> for MultiPool<E> {
+    fn from(inner: RawPool) -> Self {
+        Self { buffer_list: Vec::new(), inner }
+    }
+}
+
+impl<K: PartialEq + Clone> MultiPool<K> {
+    /// Resizes the memory pool, notifying the server the pool has changed in size.
+    ///
+    /// The wl_shm protocol only allows the pool to be made bigger. If the new size is smaller than the
+    /// current size of the pool, this function will do nothing.
+    pub fn resize(&mut self, size: usize, conn: &mut ConnectionHandle) -> io::Result<()> {
+        self.inner.resize(size, conn)
+    }
+    /// Removes the buffer with the given key from the pool and rearranges the others
+    pub fn remove(&mut self, key: &K, conn: &mut ConnectionHandle) {
+        if let Some((i, buffer)) = self.buffer_list.iter().enumerate().find(|b| b.1.key.eq(key)) {
+            let mut offset = buffer.offset;
+            self.buffer_list.remove(i);
+            for buffer_handle in &mut self.buffer_list {
+                if buffer_handle.offset > offset && buffer_handle.free.load(Ordering::Relaxed) {
```

Comment:
if the buffer is in use, then the buffer will leak.

We need a second field on the data of the buffer to tell it the pool has been destroyed and that it should release itself.

Reply:
I'm not sure about that, isn't it the same delayed reorganizing as was discussed there?

https://github.com/Smithay/client-toolkit/pull/217#discussion_r795088748